File: snow.m
#import <AppKit/AppKit.h>
#import <QuartzCore/QuartzCore.h>

@interface SnowView : NSView
@end

@implementation SnowView
- (instancetype)initWithFrame:(NSRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setWantsLayer:YES];
        self.layer.backgroundColor = [NSColor clearColor].CGColor;
        
        CAEmitterLayer *emitter = [CAEmitterLayer layer];
        emitter.emitterPosition = CGPointMake(frame.size.width / 2, frame.size.height + 10);
        emitter.emitterSize = CGSizeMake(frame.size.width, 0);
        emitter.emitterShape = kCAEmitterLayerLine;
        
        CAEmitterCell *flake = [CAEmitterCell emitterCell];
        flake.birthRate = 10.0;
        flake.lifetime = 40.0;
        flake.velocity = 40;
        flake.velocityRange = 20;
        flake.yAcceleration = -30.0;
        flake.emissionLongitude = M_PI;
        flake.emissionRange = 0.5;
        flake.scale = 0.25;
        flake.scaleRange = 0.1;
        flake.alphaSpeed = 0.0; 
        
        NSString *snowFlakeStr = @"❄";
        NSDictionary *attributes = @{ NSFontAttributeName: [NSFont systemFontOfSize:20],
                                     NSForegroundColorAttributeName: [NSColor whiteColor] };
        NSImage *img = [NSImage imageWithSize:NSMakeSize(24, 24) flipped:NO drawingHandler:^BOOL(NSRect dstRect) {
            [snowFlakeStr drawInRect:dstRect withAttributes:attributes];
            return YES;
        }];
        flake.contents = (__bridge id)[img CGImageForProposedRect:NULL context:NULL hints:NULL];
        
        emitter.emitterCells = @[flake];
        [self.layer addSublayer:emitter];
    }
    return self;
}
@end

// メニューバーを管理するためのデリゲートクラス
@interface AppDelegate : NSObject <NSApplicationDelegate>
@property (strong) NSStatusItem *statusItem;
@property (strong) NSWindow *window;
@end

@implementation AppDelegate
- (void)applicationDidFinishLaunching:(NSNotification *)notification {
    // 1. メニューバーの設定
    self.statusItem = [[NSStatusBar systemStatusBar] statusItemWithLength:NSVariableStatusItemLength];
    self.statusItem.button.title = @"❄"; // メニューバーに表示されるアイコン
    
    NSMenu *menu = [[NSMenu alloc] init];
    [menu addItemWithTitle:@"Quit" action:@selector(terminate:) keyEquivalent:@"q"];
    self.statusItem.menu = menu;

    // 2. 雪を降らせるウィンドウの設定
    NSRect screenRect = [[NSScreen mainScreen] frame];
    self.window = [[NSWindow alloc] initWithContentRect:screenRect
                                              styleMask:NSWindowStyleMaskBorderless
                                                backing:NSBackingStoreBuffered
                                                  defer:NO];
    
    [self.window setBackgroundColor:[NSColor clearColor]];
    [self.window setOpaque:NO];
    [self.window setHasShadow:NO];
    [self.window setIgnoresMouseEvents:YES];
    [self.window setLevel:NSScreenSaverWindowLevel];
    [self.window setCollectionBehavior:NSWindowCollectionBehaviorCanJoinAllSpaces | NSWindowCollectionBehaviorFullScreenAuxiliary];
    
    SnowView *view = [[SnowView alloc] initWithFrame:screenRect];
    [self.window setContentView:view];
    [self.window makeKeyAndOrderFront:nil];
}
@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSApplication *app = [NSApplication sharedApplication];
        AppDelegate *delegate = [[AppDelegate alloc] init];
        app.delegate = delegate;
        [app setActivationPolicy:NSApplicationActivationPolicyAccessory];
        [app run];
    }
    return 0;
}

File: swift/Snow.swift
import Cocoa
import QuartzCore

class SnowWindow: NSWindow {
    init(screen: NSScreen) {
        let frame = screen.frame
        super.init(contentRect: frame,
                   styleMask: .borderless,
                   backing: .buffered,
                   defer: false)
        self.isOpaque = false
        self.backgroundColor = .clear
        self.ignoresMouseEvents = true
        self.level = .screenSaver
    }
}

class SnowView: NSView {
    let emitter = CAEmitterLayer()
    
    override init(frame frameRect: NSRect) {
        super.init(frame: frameRect)
        wantsLayer = true
        
        emitter.emitterPosition = CGPoint(x: frame.width/2, y: frame.height + 10)
        emitter.emitterSize = CGSize(width: frame.width, height: 0)
        emitter.emitterShape = .line
        
        let flake = CAEmitterCell()
        flake.birthRate = 15
        flake.lifetime = 40
        flake.velocity = 40
        flake.yAcceleration = -30
        flake.emissionLongitude = .pi
        flake.emissionRange = 0.5
        flake.scale = 0.25
        
        // 雪の文字をイメージ化
        let snowStr = "❄"
        let img = NSImage(size: NSSize(width: 24, height: 24), flipped: false) { rect in
            snowStr.draw(in: rect, withAttributes: [
                .font: NSFont.systemFont(ofSize: 20),
                .foregroundColor: NSColor.white
            ])
            return true
        }
        flake.contents = img.cgImage(forProposedRect: nil, context: nil, hints: nil)
        
        emitter.emitterCells = [flake]
        layer?.addSublayer(emitter)
    }
    
    required init?(coder: NSCoder) { fatalError() }
}

let app = NSApplication.shared
app.setActivationPolicy(.accessory)

let screen = NSScreen.main!
let window = SnowWindow(screen: screen)
window.contentView = SnowView(frame: screen.frame)
window.makeKeyAndOrderFront(nil)

app.run()
